lib/Dialect/FIRRTL/Transforms/InferReadWrite.cpp

```diff
@@ -0,0 +1,268 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLAnnotations.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/Namespace.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "llvm/ADT/APSInt.h"
+#include "llvm/Support/Debug.h"
+#include <set>
+
+#define DEBUG_TYPE "firrtl-infer-read-write"
+
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+struct InferReadWritePass : public InferReadWriteBase<InferReadWritePass> {
+
+  void runOnOperation() override {
+    LLVM_DEBUG(llvm::dbgs() << "\n Running Infer Read Write on module:"
+                            << getOperation().getName());
+    ModuleNamespace modNamespace(getOperation());
+    for (MemOp mem : llvm::make_early_inc_range(
+             getOperation().getBody()->getOps<MemOp>())) {
+      size_t nReads, nWrites, nRWs;
+      mem.getNumPorts(nReads, nWrites, nRWs);
+      // Run the analysis only for Seq memories (latency=1) and a single read
+      // and write ports.
+      if (!(nReads == 1 && nWrites == 1 && nRWs == 0) ||
+          !(mem.readLatency() == 1 && mem.writeLatency() == 1))
+        continue;
+      Value rClock, wClock;
+      // The memory has exactly two ports.
+      SmallVector<Value> portTerms[2];
+      for (auto portIt : llvm::enumerate(mem.results())) {
+        // Get the port value.
+        Value portVal = portIt.value();
+        // Get the port kind.
+        bool readPort =
+            mem.getPortKind(portIt.index()) == MemOp::PortKind::Read;
+        // Iterate over all users of the port.
+        for (Operation *u : portVal.getUsers())
+          if (auto sf = dyn_cast<SubfieldOp>(u)) {
+            // Get the field name.
+            auto fName = sf.input().getType().cast<BundleType>().getElementName(
+                sf.fieldIndex());
+            // If this is the enable field, record the product terms(the And
+            // expression tree).
+            if (fName.equals("en"))
+              getProductTerms(sf, portTerms[portIt.index()]);
+            else if (fName.equals("clk")) {
+              if (readPort)
+                rClock = getConnectSrc(sf);
+              else
+                wClock = getConnectSrc(sf);
+            }
+          }
+        // End of loop for getting MemOp port users.
+      }
+      if (rClock != wClock) {
+        DenseSet<Value> rClocks, wClocks;
+        while (rClock) {
+          rClocks.insert(rClock);
+          rClock = getConnectSrc(rClock);
+        }
+
+        bool sameClock = false;
+        while (wClock) {
+          if (rClocks.find(wClock) != rClocks.end()) {
+            sameClock = true;
+            break;
+          }
+          wClock = getConnectSrc(wClock);
+        }
+        if (!sameClock)
+          continue;
+        rClock = wClock;
+      }
```

Comment:
Can you put this into a `static` or `private` utility? 

The would then enable the something like the following which is more readable:

```c++
if (!sameDriver(rClock, wClock))
  continue;
```